// cypress-controls-ext/dist/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeControls = exports.setupControlsExtensionWithEvent = exports.setupControlsExtension = exports.updateEnvVar = exports.getStoredVar = exports.setStoredVar = exports.cypressAppSelect = void 0;
/**
 * for testing purposes
 */
var common_1 = require("./common");
Object.defineProperty(exports, "cypressAppSelect", { enumerable: true, get: function () { return common_1.cypressAppSelect; } });
Object.defineProperty(exports, "setStoredVar", { enumerable: true, get: function () { return common_1.setStoredVar; } });
Object.defineProperty(exports, "getStoredVar", { enumerable: true, get: function () { return common_1.getStoredVar; } });
Object.defineProperty(exports, "updateEnvVar", { enumerable: true, get: function () { return common_1.updateEnvVar; } });
var inject_1 = require("./inject");
Object.defineProperty(exports, "setupControlsExtension", { enumerable: true, get: function () { return inject_1.setupControlsExtension; } });
Object.defineProperty(exports, "setupControlsExtensionWithEvent", { enumerable: true, get: function () { return inject_1.setupControlsExtensionWithEvent; } });
Object.defineProperty(exports, "removeControls", { enumerable: true, get: function () { return inject_1.removeControls; } });
